src/esp/sim/AbstractReplayRenderer.h

```diff
@@ -95,11 +95,16 @@ class AbstractReplayRenderer {
   void setSensorTransformsFromKeyframe(unsigned envIndex,
                                        const std::string& prefix);
 
-  // Renders and waits for the render to finish
+  // Render into the default framebuffer. Assumes the framebuffer color & depth
+  // is cleared.
+  void render();
```

Comment:
Hmm, thinking further about all this, what was wrong with calling `render(Mn::GL::defaultFramebuffer)` from the application code in the first place?

I see that it could possibly trip up [on an assert](https://github.com/facebookresearch/habitat-sim/blob/b6186aa8aaf9783f791ec4ffea00b091a766a96f/src/esp/sim/BatchReplayRenderer.cpp#L240-L242) which is there to prevent accidents like rendering to some arbitrary location but then trying to retrieve a CUDA image which would be inevitably stale or completely empty. But then the solution is to fix the root cause that causes the assert (such as not using a standalone renderer if you're rendering to a window) or rethink the use case for the assertion. Adding a completely new code path that sidesteps the assertion isn't a solution.

> Assumes the framebuffer color & depth is cleared.

Because this I see as a problem, there's a new API that *looks* easy, but it has implicit requirements that might not be obvious from reading the application code alone, and is a reason I didn't add such overload myself. Consider this, from the [replayer app](https://github.com/facebookresearch/habitat-sim/blob/b6186aa8aaf9783f791ec4ffea00b091a766a96f/src/utils/replayer/replayer.cpp#L197-L233):

```cpp
  Mn::GL::defaultFramebuffer.clear(Mn::GL::FramebufferClear::Color |
                                   Mn::GL::FramebufferClear::Depth);
                                   
  ...

  // Explicitly passing the framebuffer. The code implies it's rendering
  // to the framebuffer the application prepared above, and nothing else.
  replayRenderer_->render(Mn::GL::defaultFramebuffer);
```

Compared to 

```cpp                 
  // Renders where exactly? Is the application supposed to bind some 
  // concrete framebuffer before? Is it supposed to clear it? Or is it 
  // cleared automatically?
  replayRenderer_->render();
```

It makes things even worse / more confusing given there's a parameter-less [draw()](https://aihabitat.org/docs/habitat-sim/classesp_1_1gfx__batch_1_1RendererStandalone.html#a3e2fe82df57fbef78034b1bc7fc2a2e7) on the standalone batch renderer, which does exactly the opposite of the new `render()` here -- in the standalone batch renderer it draws into an *internal* framebuffer, clearing it before.

Reply:
I agree with those remarks. I like the explicit approach better too.

>what was wrong with calling render(Mn::GL::defaultFramebuffer) from the application code in the first place?

When writing this code, I had these two concerns:
* The assert trips-up when used from Habitat-Lab, which is a standalone renderer. Here, I am rendering to the main framebuffer, offscreen, and fetching the results using CUDA pointers.
* Magnum depedency is required by the application to pass the parameter. *This is not a valid concern anymore as we are moving forward with adding Magnum as a dependency to Habitat-Lab.*

Given this, I would roll-back the API changes, but allow for using the render function from a standalone renderer.

Reply:
> Magnum depedency is required by the application to pass the parameter.

~~If that would be a valid concern, I'd propose having an explicitly-named `renderToDefaultFramebuffer()` API on the `AbstractReplayRenderer` class instead, which calls `render(GL::defaultFramebuffer)` internally, and no other change.~~ But that still isn't the right solution I think -- see below.

> Here, I am rendering to the main framebuffer, offscreen, and fetching the results using CUDA pointers.

Okay, let's talk about how to solve this best -- does this mean you call `render(list of images)` as well as `render(default framebuffer)`? I.e., it's possible that both get called in the same frame? Um, actually, what do you call to render to just CUDA images and not fetch them via image views? I don't see any relevant API on the current `AbstractReplayRenderer` class, huh, sounds like a strange omission :thinking: 

In any case, instead of rendering twice (which also has to do all draw preparation, culling and whatnot twice) I think it'd be better to render just once the internal framebuffer and then blit it to the default one. So, adding APIs like `blitColorInto(destination, sourceRectangle, destinationRectangle)` / `blitDepthInto(destination, sourceRectangle, destinationRectangle)` to the `BatchRendererStandalone` class, and then calling those after.

Reply:
>Okay, let's talk about how to solve this best -- does this mean you call render(list of images) as well as render(default framebuffer)?

Not exactly. I'm calling *either* one or the other:

For the CUDA code path, I am using `render(default framebuffer)`, then `getCudaColorBufferDevicePointer()` to get the results. I'm using `torch::from_blob()` ([ref](https://pytorch.org/cppdocs/api/function_namespacetorch_1ad7fb2a7759ef8c9443b489ddde494787.html)) to build a tensor that references that GPU memory.

For the fallback non-CUDA code path, I am using `render(list of images)`.

>Um, actually, what do you call to render to just CUDA images and not fetch them via image views?

`getCudaColorBufferDevicePointer()` calls `RendererStandalone::colorCudaBufferDevicePointer()`, which copies the default framebuffer into a linear CUDA buffer.

Reply:
Huh...

> copies the default framebuffer

... are you sure this works and doesn't copy something stale? Because this code is always reading from the internal framebuffer, not the one that was passed to `render()`: https://github.com/facebookresearch/habitat-sim/blob/107013b63136648af00b52af189b99821a4924e9/src/esp/gfx_batch/RendererStandalone.cpp#L214-L215

I just can't see how this could do the right thing :sweat_smile: The whole `ReplayRenderer` API feels strange to me, and the CUDA pointer retrievals added in #2011 just can't work if I see correctly -- there's simply no way to render directly to the internal framebuffer and then retrieve the images as CUDA pointers. Sorry for not catching that earlier. 

I think splitting `render(list of images)` into `render()` and `colorImagesInto(list of images)` / `depthImagesInto(list of images)` could fix it, then you'll have a possibility of rendering into the internal framebuffer and then either copying them via CUDA or to the CPU. And for rendering to the default framebuffer the `blit*()` APIs I proposed above.

Reply:
I see.

Visually, the results were fine which led me to believe that this was the original intent.

>I think splitting render(list of images) into render() and colorImagesInto(list of images) / depthImagesInto(list of images) could fix it, then you'll have a possibility of rendering into the internal framebuffer and then either copying them via CUDA or to the CPU.

This makes more sense! I'll try that out.